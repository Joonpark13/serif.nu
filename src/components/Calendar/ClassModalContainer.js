import { connect } from 'react-redux';
import toJS from 'util/to-js';
import { removeSection } from 'actions';
import ClassModal from './ClassModal';

/* istanbul ignore next */
function mapDispatchToProps(dispatch) {
  return {
    removeSection: (sectionId, sectionColor) => dispatch(removeSection(sectionId, sectionColor)),
  };
}

const ClassModalContainer = connect(null, mapDispatchToProps)(toJS(ClassModal));

export default ClassModalContainer;
