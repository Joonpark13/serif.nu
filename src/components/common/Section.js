import React from 'react';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';

function Section({ classes, leftHeaderContent, rightHeaderContent, sectionName, onClick }) {
  return (
    <Paper elevation={2} className={classes.paper} onClick={onClick}>
      <div className={classes.container}>
        <div className={classes.header}>
          <Typography variant="caption" className={classes.text}>
            {leftHeaderContent}
          </Typography>

          <Typography variant="caption" className={classes.text}>
            {rightHeaderContent}
          </Typography>
        </div>

        <Typography variant="subtitle2" className={`${classes.text} ${classes.name}`} noWrap>
          {sectionName}
        </Typography>
      </div>
    </Paper>
  );
}

Section.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
    container: PropTypes.string,
    header: PropTypes.string,
    text: PropTypes.string,
    name: PropTypes.string,
  }).isRequired,
  leftHeaderContent: PropTypes.string,
  rightHeaderContent: PropTypes.string,
  sectionName: PropTypes.string,
  onClick: PropTypes.func,
};

Section.defaultProps = {
  leftHeaderContent: '',
  rightHeaderContent: '',
  sectionName: '',
  onClick: () => {},
};

export default Section;
