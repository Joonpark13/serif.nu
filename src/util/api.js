/* eslint-disable import/prefer-default-export */
export function searchURL(query) {
  return `${process.env.API_URL}/search?term_id=4720&search_query=${query}`;
}

export function sectionsURL(termId, schoolAbbv, subjectAbbv, courseAbbv) {
  return `${process.env.API_URL}/sections?term_id=${termId}&school_abbv=${schoolAbbv}&subject_abbv=${subjectAbbv}&course_abbv=${courseAbbv}`;
}
