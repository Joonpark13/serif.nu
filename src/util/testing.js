/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import * as reactRedux from 'react-redux';
import { shallow } from 'enzyme';
import * as useSelector from './use-selector';

jest.mock('./use-selector');

// Helper function that returns a function which creates enzyme wrappers with default props
export function wrapperCreator(Component, defaultProps = {}) {
  let classes;
  return props => shallow(<Component classes={classes} {...defaultProps} {...props} />);
}

/* istanbul ignore next */
export function mockUseSelector(...args) {
  useSelector.default.mockReset();
  args.forEach(arg => useSelector.default.mockReturnValueOnce(arg));

  return useSelector.default;
}

/* istanbul ignore next */
export function mockUseDispatch() {
  const dispatchSpy = jest.fn();
  reactRedux.useDispatch.mockReturnValue(dispatchSpy);
  return dispatchSpy;
}
