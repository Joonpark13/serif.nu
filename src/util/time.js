export function formatMinute(minute) {
  if (minute < 10) return `0${minute}`;
  return `${minute}`;
}

export function formatTime(time, includeAmPm = true) {
  if (time.hour === 0) return `12:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour < 12) return `${time.hour}:${formatMinute(time.minute)}${includeAmPm ? ' AM' : ''}`;
  if (time.hour === 12) return `12:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
  return `${time.hour - 12}:${formatMinute(time.minute)}${includeAmPm ? ' PM' : ''}`;
}

export function getFormattedClassEvent(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  return `${formatTime(event.start, false)} - ${formatTime(event.end, false)}`;
}

export function getFormattedClassSchedule(event) {
  if (event.dow === 'TBA' || event.start === 'TBA' || event.end === 'TBA') {
    return 'TBA';
  }
  const dowStr = event.dow.join('');
  return `${dowStr} ${formatTime(event.start, true)} - ${formatTime(event.end, true)}`;
}

export function getDurationInHours(event) {
  const hoursDiff = event.end.hour - event.start.hour;
  const minutesDiff = event.end.minute - event.start.minute;
  return hoursDiff + minutesDiff / 60;
}

export function isBefore(time1, time2, orEqual = false) {
  if (time1.hour < time2.hour) {
    return true;
  }
  if (time1.hour === time2.hour) {
    if (orEqual) return time1.minute <= time2.minute;
    return time1.minute < time2.minute;
  }
  return false;
}

export function overlaps(event1, event2) {
  if (event1.dow !== event2.dow) {
    return false;
  }
  return !(isBefore(event1.end, event2.start) || isBefore(event2.end, event1.start));
}
