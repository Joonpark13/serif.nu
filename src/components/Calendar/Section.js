import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import { getDurationInHours, getFormattedClassEvent } from 'util/time';

export const MAX_WIDTH_PERCENT = 97;

export const styles = {
  paper: {
    position: 'absolute',
    top: ({ section }) => {
      const minute = section.event.start.minute;
      const offset = Math.round(minute / 60 * 100);
      return `${offset}%`;
    },
    left: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth * section.column}%`,
    height: ({ section }) => {
      const durationInHours = getDurationInHours(section.event);
      const heightInPercent = Math.round(durationInHours * 100);
      return `${heightInPercent}%`;
    },
    width: ({ section }) => `${MAX_WIDTH_PERCENT * section.columnWidth}%`,
    backgroundColor: ({ section }) => section.color,
    overflow: 'hidden',
  },
  container: {
    margin: '3px',
  },
  text: {
    color: 'white',
  },
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
  },
};

function Section({ classes, section }) {
  return (
    <Paper className={classes.paper}>
      <div className={classes.container}>
        <div className={classes.header}>
          <Typography variant="caption" className={classes.text}>
            {getFormattedClassEvent(section.event)}
          </Typography>

          <Typography variant="caption" className={classes.text}>
            {`${section.subjectId} ${section.courseId}`}
          </Typography>
        </div>

        <Typography variant="subtitle2" className={`${classes.text} ${classes.name}`} noWrap>
          {section.name}
        </Typography>
      </div>
    </Paper>
  );
}

Section.propTypes = {
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export { Section as UnstyledSection };
export default withStyles(styles)(Section);
