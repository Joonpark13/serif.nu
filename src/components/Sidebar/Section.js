import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import { getFormattedClassSchedule } from 'util/time';

export const styles = {
  sectionTitle: {
    fontWeight: 'bold',
  },
};

function Section({ addSection, section, disabled, classes }) {
  return (
    <ListItem
      key={section.id}
      button
      onClick={() => addSection(section)}
      disabled={disabled}
    >
      <ListItemText>
        <Typography variant="h6" className={classes.sectionTitle}>
          {`Section ${section.sectionNumber}`}
        </Typography>

        {section.topic && <Typography>{section.topic}</Typography>}

        {section.schedule.map((scheduleObj, index) => (
          /* eslint-disable react/no-array-index-key */
          <Typography key={index}>
            {getFormattedClassSchedule(scheduleObj)}
          </Typography>
          /* eslint-enable react/no-array-index-key */
        ))}

        <Typography>{section.schedule[0].location}</Typography>

        <Typography>
          {section.instructors.map((teacher, idx) => (
            `${teacher}${idx !== section.instructors.length - 1 ? ', ' : ''}`))
          }
        </Typography>
      </ListItemText>
    </ListItem>
  );
}

Section.propTypes = {
  addSection: PropTypes.func.isRequired,
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  disabled: PropTypes.bool,
};

Section.defaultProps = {
  disabled: false,
};

export { Section as UnstyledSection };
export default withStyles(styles)(Section);
