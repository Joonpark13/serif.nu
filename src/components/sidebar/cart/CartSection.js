import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import { getFormattedClassSchedule } from 'util/time';
import Section from 'components/common/Section';
import ClassModalContainer from 'components/calendar/ClassModalContainer';

export const styles = {
  paper: {
    backgroundColor: ({ section }) => section.color,
    overflow: 'hidden',
    margin: '10px',
    cursor: 'pointer',
  },
  container: {
    margin: '3px',
  },
  text: {
    color: 'white',
  },
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
  },
};

class CartSection extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showDialog: false,
    };
    this.toggleDialog = this.toggleDialog.bind(this);
  }

  toggleDialog() {
    this.setState(state => ({ showDialog: !state.showDialog }));
  }

  render() {
    const { classes, section } = this.props;
    const { showDialog } = this.state;

    let leftHeaderContent;
    if (section.schedule.length > 1) {
      leftHeaderContent = section.schedule
        .map(schedule => getFormattedClassSchedule(schedule))
        .reduce(
          (resultStr, formattedStr) => `${resultStr}, ${formattedStr}`,
        );
    } else {
      leftHeaderContent = getFormattedClassSchedule(section.schedule[0]);
    }
    const rightHeaderContent = `${section.subjectId} ${section.courseId}`;

    return (
      <React.Fragment>
        <Section
          onClick={this.toggleDialog}
          classes={classes}
          leftHeaderContent={leftHeaderContent}
          rightHeaderContent={rightHeaderContent}
          sectionName={section.name}
        />
        <ClassModalContainer
          section={section}
          showDialog={showDialog}
          toggleDialog={this.toggleDialog}
        />
      </React.Fragment>
    );
  }
}

CartSection.propTypes = {
  section: PropTypes.objectOf(PropTypes.any).isRequired, // TODO
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
};

export { CartSection as UnstyledCartSection };
export default withStyles(styles)(CartSection);
