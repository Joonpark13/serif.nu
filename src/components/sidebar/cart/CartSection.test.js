import React from 'react';
import { shallow } from 'enzyme';
import { mockStyles } from 'util/testing';
import * as timeUtils from 'util/time';
import Section from 'components/common/Section';
import ClassModalContainer from 'components/calendar/ClassModalContainer';
import { UnstyledCartSection, styles } from './CartSection';

describe('CartSection', () => {
  beforeEach(() => {
    timeUtils.getFormattedClassSchedule = jest.fn();
    timeUtils.getFormattedClassSchedule.mockReturnValue('schedules');
  });

  describe('dynamic styles', () => {
    const color = 'some color';
    const section = { color };
    it('correctly grabs the section background color', () => {
      expect(styles.paper.backgroundColor({ section })).toBe(color);
    });
  });

  it('renders correctly', () => {
    const classes = mockStyles(styles);
    const testSection = { schedules: [{}], subjectId: 'EECS', courseId: '111-0' };
    const wrapper = shallow(
      <UnstyledCartSection section={testSection} classes={classes} />,
    );

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('formats left header content correctly when multiple schedules present', () => {
    const classes = mockStyles(styles);
    const testSection = { schedules: [{}, {}], subjectId: 'EECS', courseId: '111-0' };
    const wrapper = shallow(
      <UnstyledCartSection section={testSection} classes={classes} />,
    );

    expect(wrapper.find(Section).prop('leftHeaderContent')).toBe('schedules, schedules');
  });

  it('opens modal on click', () => {
    const classes = mockStyles(styles);
    const testSection = { schedules: [{}], subjectId: 'EECS', courseId: '111-0' };
    const wrapper = shallow(
      <UnstyledCartSection section={testSection} classes={classes} />,
    );

    wrapper.find(Section).simulate('click');

    expect(wrapper.find(ClassModalContainer).prop('showDialog')).toBe(true);
  });
});
