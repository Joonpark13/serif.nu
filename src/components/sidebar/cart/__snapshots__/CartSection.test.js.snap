// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartSection renders correctly 1`] = `
<React.Fragment>
  <Section
    classes={
      Object {
        "container": "",
        "header": "",
        "paper": "",
        "text": "",
      }
    }
    leftHeaderContent="schedules"
    onClick={[Function]}
    rightHeaderContent="EECS 111-0"
    sectionName=""
  />
  <Connect(Component)
    section={
      Object {
        "courseId": "111-0",
        "schedules": Array [
          Object {},
        ],
        "subjectId": "EECS",
      }
    }
    showDialog={false}
    toggleDialog={[Function]}
  />
</React.Fragment>
`;
